### PR TITLE
additional Subaru fwVersion

### PR DESCRIPTION
**Pull request by spawahh**
Description: [Added additional 'fwVersion' for Subaru Crosstrek]
Route: [a3066dcab487207a|2023-08-12--15-12-44]

### CODE REVIEW

File: selfdrive/car/subaru/values.py [hunk 1]
```diff
@@ -262,19 +262,22 @@ def init_make(self, CP: car.CarParams):
       b'\xa2 !`\000',
       b'\xf1\x00\xb2\x06\x04',
       b'\xa2  `\x00',
+      b'\xa2 !i\x00',
```
Comment (jnewb1) [inline selfdrive/car/subaru/values.py:265]: duplicate

File: selfdrive/car/subaru/values.py [hunk 2]
```diff
@@ -262,19 +262,22 @@ def init_make(self, CP: car.CarParams):
...
     ],
     (Ecu.eps, 0x746, None): [
       b'\x9a\xc0\000\000',
       b'\n\xc0\004\000',
       b'\x9a\xc0\x04\x00',
       b'\n\xc0\x04\x01',
+      b'\n\xc0\x04\x00',
```
Comment (jnewb1) [inline selfdrive/car/subaru/values.py:272]: duplicate

File: selfdrive/car/subaru/values.py [hunk 3]
```diff
@@ -299,6 +303,10 @@ def init_make(self, CP: car.CarParams):
       b'\xe9\xf6F0\x00',
       b'\xe9\xf5B0\x00',
       b'\xe9\xf6B0\x00',
+      b'\xf1\x82\xe9\xf6B0\x00',
+    ],
+    (Ecu.fwdRadar, 0x7d0, None): [
```
Comment (jnewb1) [inline selfdrive/car/subaru/values.py:308]: there is no fwd radar on subaru. you are probably mistaking it for a different ECU address on another brand

Comment (spawahh) [reply selfdrive/car/subaru/values.py:308]: I'm sorry for the duplicates, i try to double check but somehow the search only works half the time... Yeah, I was just pulling all the carParams ECUs and it was in there...

**Review by spawahh — COMMENTED** [review]
a3066dcab487207a|2023-08-12--16-15-12

**Review by spawahh — COMMENTED** [review]
a3066dcab487207a|2023-08-12--16-15-12

**Comment by jnewb1** [comment]
I will merge the engine fw because it's new.

but can you explain where you are getting these from?:

transmission: \xf1\x82\xe9\xf6B0\x00
camera: \xf1\x00\xac\x03\x00

it looks like part of the UDS request is in these

**Review by spawahh — COMMENTED** [review]
unable to confirm

**Comment by spawahh** [comment]
Original firmware retrieved from: a3066dcab487207a|2023-08-12--20-49-49
Test of 80c48d3: a3066dcab487207a|2023-08-12--22-11-10--0

**Comment by spawahh** [comment]
{'carParams': {'alternativeExperience': 33,
               'autoResumeSng': True,
               'carFingerprint': 'mock',
               'carFw': [{'address': 2017,
                          'brand': 'hyundai',
                          'bus': 1,
                          'ecu': 'transmission',
                          'fwVersion': b'\xf1\x00\xd7\x10@',
                          'logging': False,
                          'obdMultiplexing': True,
                          'request': [b'"\xf1\x00'],
                          'responseAddress': 2025,
                          'subAddress': 0},
                         {'address': 1988,
                          'brand': 'hyundai',
                          'bus': 1,
                          'ecu': 'fwdCamera',
                          'fwVersion': b'\xf1\x00\xac\x03\x00',
                          'logging': False,
                          'obdMultiplexing': True,
                          'request': [b'"\xf1\x00'],
                          'responseAddress': 1996,
                          'subAddress': 0},
                         {'address': 2000,
                          'brand': 'hyundai',
                          'bus': 1,
                          'ecu': 'fwdRadar',
                          'fwVersion': b'\xf1\x00\x00\x00\x02',
                          'logging': False,
                          'obdMultiplexing': True,
                          'request': [b'"\xf1\x00'],
                          'responseAddress': 2008,
                          'subAddress': 0},
                         {'address': 2000,
                          'brand': 'hyundai',
                          'bus': 1,
                          'ecu': 'fwdRadar',
                          'fwVersion': b'\xf1\x00\x00\x00\x02',
                          'logging': False,
                          'obdMultiplexing': True,
                          'request': [b'"\xf1\x87\xf1\x81\xf1\x00'],
                          'responseAddress': 2008,
                          'subAddress': 0},
                         {'address': 2017,
                          'brand': 'hyundai',
                          'bus': 1,
                          'ecu': 'transmission',
                          'fwVersion': b'\xf1\x00\xd7\x10@',
                          'logging': False,
                          'obdMultiplexing': True,
                          'request': [b'"\xf1\x87\xf1\x81\xf1\x00'],
                          'responseAddress': 2025,
                          'subAddress': 0},
                         {'address': 1988,
                          'brand': 'hyundai',
                          'bus': 0,
                          'ecu': 'fwdCamera',
                          'fwVersion': b'\xf1\x00\xac\x03\x00',
                          'logging': False,
                          'obdMultiplexing': True,
                          'request': [b'"\xf1\x00'],
                          'responseAddress': 1996,
                          'subAddress': 0},
                         {'address': 1862,
                          'brand': 'subaru',
                          'bus': 1,
                          'ecu': 'eps',
                          'fwVersion': b'\n\xc0\x04\x00',
                          'logging': False,
                          'obdMultiplexing': True,
                          'request': [b'>\x00', b'"\xf1\x82'],
                          'responseAddress': 1870,
                          'subAddress': 0},
                         {'address': 2016,
                          'brand': 'subaru',
                          'bus': 1,
                          'ecu': 'engine',
                          'fwVersion': b'\xe6"fp\x07',
                          'logging': False,
                          'obdMultiplexing': True,
                          'request': [b'>\x00', b'"\xf1\x82'],
                          'responseAddress': 2024,
                          'subAddress': 0},
                         {'address': 2017,
                          'brand': 'subaru',
                          'bus': 1,
                          'ecu': 'transmission',
                          'fwVersion': b'\xe9\xf6B0\x00',
                          'logging': False,
                          'obdMultiplexing': True,
                          'request': [b'>\x00', b'"\xf1\x82'],
                          'responseAddress': 2025,
                          'subAddress': 0},
                         {'address': 1968,
                          'brand': 'subaru',
                          'bus': 1,
                          'ecu': 'abs',
                          'fwVersion': b'\xa2 !i\x00',
                          'logging': False,
                          'obdMultiplexing': True,
                          'request': [b'>\x00', b'"\xf1\x82'],
                          'responseAddress': 1976,
                          'subAddress': 0},
                         {'address': 1927,
                          'brand': 'subaru',
                          'bus': 1,
                          'ecu': 'fwdCamera',
                          'fwVersion': b'\x00\x00e\x8f\x1f@ )',
                          'logging': False,
                          'obdMultiplexing': True,
                          'request': [b'>\x00', b'"\xf1\x82'],
                          'responseAddress': 1935,
                          'subAddress': 0},
                         {'address': 2017,
                          'brand': 'volkswagen',
                          'bus': 1,
                          'ecu': 'transmission',
                          'fwVersion': b'\xf1\x82\xe9\xf6B0\x00',
                          'logging': False,
                          'obdMultiplexing': True,
                          'request': [b'"\xf1\x87\xf1\x89\xf1\x82'],
                          'responseAddress': 2025,
                          'subAddress': 0}],

**Comment by jnewb1** [comment]
> {'carParams': {'alternativeExperience': 33,
> 
>                'autoResumeSng': True,
> 
>                'carFingerprint': 'mock',
> 
>                'carFw': [{'address': 2017,
> 
>                           'brand': 'hyundai',
> 
>                           'bus': 1,
> 
>                           'ecu': 'transmission',
> 
>                           'fwVersion': b'\xf1\x00\xd7\x10@',
> 
>                           'logging': False,
> 
>                           'obdMultiplexing': True,
> 
>                           'request': [b'"\xf1\x00'],
> 
>                           'responseAddress': 2025,
> 
>                           'subAddress': 0},
> 
>                          {'address': 1988,
> 
>                           'brand': 'hyundai',
> 
>                           'bus': 1,
> 
>                           'ecu': 'fwdCamera',
> 
>                           'fwVersion': b'\xf1\x00\xac\x03\x00',
> 
>                           'logging': False,
> 
>                           'obdMultiplexing': True,
> 
>                           'request': [b'"\xf1\x00'],
> 
>                           'responseAddress': 1996,
> 
>                           'subAddress': 0},
> 
>                          {'address': 2000,
> 
>                           'brand': 'hyundai',
> 
>                           'bus': 1,
> 
>                           'ecu': 'fwdRadar',
> 
>                           'fwVersion': b'\xf1\x00\x00\x00\x02',
> 
>                           'logging': False,
> 
>                           'obdMultiplexing': True,
> 
>                           'request': [b'"\xf1\x00'],
> 
>                           'responseAddress': 2008,
> 
>                           'subAddress': 0},
> 
>                          {'address': 2000,
> 
>                           'brand': 'hyundai',
> 
>                           'bus': 1,
> 
>                           'ecu': 'fwdRadar',
> 
>                           'fwVersion': b'\xf1\x00\x00\x00\x02',
> 
>                           'logging': False,
> 
>                           'obdMultiplexing': True,
> 
>                           'request': [b'"\xf1\x87\xf1\x81\xf1\x00'],
> 
>                           'responseAddress': 2008,
> 
>                           'subAddress': 0},
> 
>                          {'address': 2017,
> 
>                           'brand': 'hyundai',
> 
>                           'bus': 1,
> 
>                           'ecu': 'transmission',
> 
>                           'fwVersion': b'\xf1\x00\xd7\x10@',
> 
>                           'logging': False,
> 
>                           'obdMultiplexing': True,
> 
>                           'request': [b'"\xf1\x87\xf1\x81\xf1\x00'],
> 
>                           'responseAddress': 2025,
> 
>                           'subAddress': 0},
> 
>                          {'address': 1988,
> 
>                           'brand': 'hyundai',
> 
>                           'bus': 0,
> 
>                           'ecu': 'fwdCamera',
> 
>                           'fwVersion': b'\xf1\x00\xac\x03\x00',
> 
>                           'logging': False,
> 
>                           'obdMultiplexing': True,
> 
>                           'request': [b'"\xf1\x00'],
> 
>                           'responseAddress': 1996,
> 
>                           'subAddress': 0},
> 
>                          {'address': 1862,
> 
>                           'brand': 'subaru',
> 
>                           'bus': 1,
> 
>                           'ecu': 'eps',
> 
>                           'fwVersion': b'\n\xc0\x04\x00',
> 
>                           'logging': False,
> 
>                           'obdMultiplexing': True,
> 
>                           'request': [b'>\x00', b'"\xf1\x82'],
> 
>                           'responseAddress': 1870,
> 
>                           'subAddress': 0},
> 
>                          {'address': 2016,
> 
>                           'brand': 'subaru',
> 
>                           'bus': 1,
> 
>                           'ecu': 'engine',
> 
>                           'fwVersion': b'\xe6"fp\x07',
> 
>                           'logging': False,
> 
>                           'obdMultiplexing': True,
> 
>                           'request': [b'>\x00', b'"\xf1\x82'],
> 
>                           'responseAddress': 2024,
> 
>                           'subAddress': 0},
> 
>                          {'address': 2017,
> 
>                           'brand': 'subaru',
> 
>                           'bus': 1,
> 
>                           'ecu': 'transmission',
> 
>                           'fwVersion': b'\xe9\xf6B0\x00',
> 
>                           'logging': False,
> 
>                           'obdMultiplexing': True,
> 
>                           'request': [b'>\x00', b'"\xf1\x82'],
> 
>                           'responseAddress': 2025,
> 
>                           'subAddress': 0},
> 
>                          {'address': 1968,
> 
>                           'brand': 'subaru',
> 
>                           'bus': 1,
> 
>                           'ecu': 'abs',
> 
>                           'fwVersion': b'\xa2 !i\x00',
> 
>                           'logging': False,
> 
>                           'obdMultiplexing': True,
> 
>                           'request': [b'>\x00', b'"\xf1\x82'],
> 
>                           'responseAddress': 1976,
> 
>                           'subAddress': 0},
> 
>                          {'address': 1927,
> 
>                           'brand': 'subaru',
> 
>                           'bus': 1,
> 
>                           'ecu': 'fwdCamera',
> 
>                           'fwVersion': b'\x00\x00e\x8f\x1f@ )',
> 
>                           'logging': False,
> 
>                           'obdMultiplexing': True,
> 
>                           'request': [b'>\x00', b'"\xf1\x82'],
> 
>                           'responseAddress': 1935,
> 
>                           'subAddress': 0},
> 
>                          {'address': 2017,
> 
>                           'brand': 'volkswagen',
> 
>                           'bus': 1,
> 
>                           'ecu': 'transmission',
> 
>                           'fwVersion': b'\xf1\x82\xe9\xf6B0\x00',
> 
>                           'logging': False,
> 
>                           'obdMultiplexing': True,
> 
>                           'request': [b'"\xf1\x87\xf1\x89\xf1\x82'],
> 
>                           'responseAddress': 2025,
> 
>                           'subAddress': 0}],

If the "brand" isn't "Subaru", it means you are looking at the wrong ECU

**Pull request merged**